torch/fx/experimental/symbolic_shapes.py

```diff
@@ -254,10 +254,50 @@ def _sympystr(self, printer):
 
         @classmethod
         def eval(cls, base, divisor):
+            # NOTE [ Checking types with SymPy ]
+            # Python has a dedicated complex type, but SymPy represents complex
+            # with Add exprs.
+            #
+            # isinstance doesn't work for arbitrary exprs and for Symbols, even
+            # when the latter have integer=True set during construction. So you
+            # are supposed to use the is_* properties, but it's also not always
+            # reliable as 0 has both is_integer and is_real set to True no
+            # matter whether it's constructed as a Float or an Integer.
+            #
+            # Also, booleans cannot be used in arithmetic exprs in SymPy and
+            # cannot be passed to the Integer and Float constructors.
+            # https://github.com/pytorch/pytorch/issues/90900
+            def check_supported_type(x):
+                # Note: we disallow complex as in regular Python, but we have to
+                # check for is_integer and is_real explicitly because Integer
+                # and Float have is_complex set to True. We also don't allow
+                # booleans because they cannot be used in arithmetic exprs in
+                # SymPy.
+                if (not x.is_integer and not x.is_real and x.is_complex
+                        or x.is_Boolean):
```

Comment:
I have trouble telling whether or not and/or have higher precedence, so best to parenthesize more here (in contrast, the outer parenthesis is completely unnecessary)

Reply:
See the third bullet point here https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval. I recommend not failing here if assumptions are None.